src/keepass2john.c

```diff
@@ -857,19 +857,51 @@ static void process_database(char* encryptedDatabase)
 
 		/*
 		 * As in Keepass 2.x implementation:
-		 *  if keyfile is an xml, get <Data> content
+		 *  if keyfile is Version 1 XML, get <Data> content in Base64
+		 *  if keyfile is Version 2 XML, get <Data> content in hex
 		 *  if filesize_keyfile == 32 then assume byte_array
 		 *  if filesize_keyfile == 64 then assume hex(byte_array)
 		 *  else byte_array = sha256(keyfile_content)
 		 */
+		buffer[filesize_keyfile] = 0;
 
-		if (!memcmp((char*) buffer, "<?xml", 5) &&
-		    ((p = strstr((char*) buffer, "<Key>")) != NULL) &&
+		if (!strncmp(buffer, "<?xml", 5) &&
+		    ((p = strstr(buffer, "<Version>1.0")) != NULL) &&
+		    ((p = strstr(buffer, "<Key>")) != NULL) &&
 		    ((p = strstr(p, "<Data>")) != NULL)) {
 			p += strlen("<Data>");
+			while (*p == ' ' || *p == '\t' || *p == '\r' || *p == '\n')
+				p++;
 			data = p;
 			p = strstr(p, "</Data>");
-			printf ("%s", base64_convert_cp(data, e_b64_mime, p - data, b64_decoded, e_b64_hex, sizeof(b64_decoded), flg_Base64_NO_FLAGS, 0));
+			if (p == NULL || p - buffer < 44) {
```

Comment:
Should probably be `!= 44` rather than `< 44` here. That would be consistent with the check for hex.